File: frontend/src/elements/ai/aiElement.ts
import type { NodeElementDefinition } from '../types';
import AIEditor from './AIEditor';
import { baseNodeConfig } from '../shared/baseNodeConfig';
import { outputFormatContext } from '../shared/generationContext';

export const aiElement: NodeElementDefinition = {
  nodeType: 'ai',
  authoredFile: () => ({ extension: '.md', what: 'this system prompt' }),
  generation: {
    // The one element whose request lives on the node rather than in its config.
    promptField: 'description',
    targetField: 'system_prompt',
    guard: 'Please add a description first.',
    success: '✅ Prompt generated!',
    context: (node) => outputFormatContext(node.config),
  },
  describeOutput: (node) => {
    const format = node.config.output_format;
    if (!format || format === 'text') return 'text';
    const detail = format === 'custom' && node.config.output_format_prompt
      ? `: ${node.config.output_format_prompt}` : '';
    return `${format}${detail}`;
  },
  outputContract: 'format',
  ConfigEditor: AIEditor,
  create: (id) => ({
    id,
    node_type: 'ai',
    label: 'AI Node',
    description: 'Send a prompt to an AI model',
    position: { x: 0, y: 0 },
    inputs: [
      { id: 'prompt', name: 'Prompt batch', kind: 'input', data_type: 'text', multi: true, required: true, description: 'One prompt per batch item' },
      { id: 'context', name: 'Context', kind: 'input', data_type: 'any', multi: true, required: false, description: 'Additional context' },
    ],
    outputs: [{ id: 'output', name: 'Output batch', kind: 'output', data_type: 'text', multi: true, required: false, description: 'One response per prompt item' }],
    config: { ...baseNodeConfig(), system_prompt: 'You are a helpful assistant.' },
  }),
};

File: backend/app/elements/ai/ai_element.py
"""The `ai` node element: run inputs through a configured LLM completion call."""

from __future__ import annotations

from typing import Optional

from app.elements.base import AuthoredFile, Generation, NodeElement
from app.models.graph import GraphNode, NodeType
from app.services import file_service

_FORMAT_LABELS = {
    "json": "a JSON object or array",
    "csv": "CSV with a header row",
    "csv_list": "CSV rows as a list of lists",
}




def _as_image_url(value) -> Optional[str]:
    """
    *value* as something a vision model can be sent, or None if it is not an
    image at all -- in which case it stays ordinary prompt text.
    """
    if not isinstance(value, str):
        return None
    if value.startswith("data:image/"):
        return value
    if file_service.is_image_path(value):
        try:
            return file_service.image_data_url(value)
        except (OSError, ValueError):
            return None
    return None


class AIElement(NodeElement):
    node_type = NodeType.AI
    config_fields = (
        "system_prompt", "ai_provider", "ai_model", "temperature", "send_images",
        "output_format", "output_format_prompt",
        "batch_mode", "batch_concurrency", "read_file_inputs",
    )


    def generation(self) -> Generation:
        """The system prompt, written from the node's own description -- the one
        element whose request lives on the node rather than in its config."""
        return Generation(
            kind="prompt", prompt_field="description", target_field="system_prompt",
            prompt_on_node=True,
            guard="Please add a description first.",
            success="✅ Prompt generated!",
        )

    def authored_file(self, node: GraphNode) -> AuthoredFile:
        """The system prompt -- the thing actually written for an ai node.

        Markdown, not a script that calls the model: such a script would be a
        second implementation of what ai_service already does, and would drift
        from it the moment either changed.
        """
        return AuthoredFile(
            body_field="system_prompt", prompt_field="description",
            extension=".md", prompt_on_node=True,
        )
